pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,463 @@
+	// 2. We created an entry for the node but have not opened a connection.
+	// But if we added the node via Connect, then it should open the connection.
+	// It doesn't make sense to open a new one here.
+	s, info, err := t.getNodeStreamAndInfo(dest)
+	if err != nil {
+		return err
+	}
+
+	outBytes, err := proto.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("failed to marshal message: %w", err)
+	}
+
+	msg := TransportMessage{t.ownID.Pb(), outBytes}
+	buf := new(bytes.Buffer)
+	err = msg.MarshalCBOR(buf)
+	if err != nil {
+		return fmt.Errorf("failed to CBOR marshal message: %w", err)
+	}
+
+	if ctx.Err() != nil {
+		return err
+	}
+
+	w := bufio.NewWriter(s)
+	_, err = w.Write(buf.Bytes())
+	if err == nil {
+		err = w.Flush()
+	}
+	// If we cannot write to the stream then close the connection and reconnect.
+	if err != nil {
+		if err := t.host.Network().ClosePeer(info.ID); err != nil {
+			t.logger.Log(logging.LevelError, "could not close the connection to node", "src", t.ownID, "dst", dest, "err", err)
+		}
+		t.connWg.Add(1)
+		go t.connectToNode(ctx, dest, t.connWg)
+		return errors.Wrapf(err, "%s failed to send data to %s", t.ownID, dest)
+	}
+
+	return nil
+}
+
+func (t *Transport) mirHandler(s network.Stream) {
+	t.logger.Log(logging.LevelDebug, "mir handler started", "src", t.ownID, "dst", s.ID())
+	defer t.logger.Log(logging.LevelDebug, "mir handler stopped", "src", t.ownID, "dst", s.ID())
+
+	defer func() {
+		t.logger.Log(logging.LevelDebug, "mir handler is closing stream", "src", t.ownID, "dst", s.ID())
+		err := s.Close()
+		if err != nil {
+			t.logger.Log(logging.LevelError, "closing stream", "src", t.ownID, "dst", s.ID(), "err", err)
+		}
+	}() // nolint
+
+	for {
+		var msg TransportMessage
+		err := msg.UnmarshalCBOR(s)
+		if err != nil {
+			t.logger.Log(logging.LevelError, "failed to read mir request", "src", t.ownID, "dst", s.ID(), "err", err)
+			return
+		}
+
+		var payload messagepb.Message
+
+		if err := proto.Unmarshal(msg.Payload, &payload); err != nil {
+			t.logger.Log(logging.LevelError, "failed to unmarshall mir request", "src", t.ownID, "dst", s.ID(), "err", err)
+			return
+		}
+
+		t.incomingMessages <- events.ListOf(
+			events.MessageReceived(types.ModuleID(payload.DestModule), types.NodeID(msg.Sender), &payload),
+		)
+	}
+}
+
+func (t *Transport) addOutboundStream(nodeID types.NodeID, s network.Stream) {
+	t.nodesLock.Lock()
+	defer t.nodesLock.Unlock()
+
+	node, found := t.nodes[nodeID]
+	if !found {
+		t.logger.Log(logging.LevelWarn, "addOutboundStream: failed to find the node", "src", t.ownID, "node", nodeID)
+		return
+	}
+	node.Stream = s
+}
+
+func (t *Transport) streamExists(nodeID types.NodeID) bool {
+	t.nodesLock.RLock()
+	defer t.nodesLock.RUnlock()
+
+	v, found := t.nodes[nodeID]
+	return found && v.Stream != nil
+}
+
+func (t *Transport) getNodeStreamAndInfo(nodeID types.NodeID) (network.Stream, *peer.AddrInfo, error) {
+	t.nodesLock.Lock()
```

Comment:
This is not enough, you need to keep the lock while you write the stream. I would:
- Make this function `getNodeStreamAndInfoWithoutLock`.
- Use this function when getting the stream to write by locking before getting and releasing after flush.
- Create a `getNodeStreamAndInfoWithLock` and keep the `RLock` you had before.

Reference: https://github.com/filecoin-project/mir/pull/230/commits/1667e25c53c50b79e0ab4c9d2a9bb9ba5b4284a0

Reply:
Fixed